lib/puttyhosts/puttyhosts.go

```diff
@@ -0,0 +1,132 @@
+//     trusted hostnames for each to be added manually by end users as part of session configuration.
+//   - This process is mandatory for validation of host CAs in PuTTY to work, but is a cumbersome manual process with many
+//     clicks required in a nested interface. Instead, this function is called as part of `tsh puttyconfig` to examine the
+//     existing list of trusted hostnames and automate the process of adding a new valid hostname to a given host CA.
+//
+// Connection flow:
+//   - When connecting to a host which presents a host CA, PuTTY searches its list of CAs to find any which are considered
+//     valid for that hostname, then checks whether the host's presented CA matches any of them. If there is a CA -> hostname
+//     match, the connection will continue successfully. If not, an error will be shown.
+//
+// Intended operation of this function:
+//   - This function is passed the current list of trusted hostnames for a given host CA (retrieved from the registry), along
+//     with a new hostname entry (from tsh puttyconfig <hostname>) which should be added to the list.
+//   - It appends the new hostname to the end of the hostList
+//   - All hostnames in the hostList are converted to their wildcard form if they contain a dot (test.example.com -> *.example.com)
+//     and are grouped together.
+//   - If a wildcard group only contains a single hostname which would be matched by its wildcard equivalent, that hostname is added
+//     to the hostList verbatim to prevent inadvertently matching against too many hosts with the same wildcard.
+//   - If a wildcard matches more than one hostname, the wildcard will be added to the hostList instead and the single hostnames
+//     discarded.
+//   - The hostList is then sorted alphabetically and returned.
+//
+// This is an effort to keep the length of hostList as short as possible for efficiency and tidiness, while not using any more
+// wildcards than necessary and preventing the need for end users to manually configure their trusted host CAs.
+func AddHostToHostList(hostList []string, hostname string) []string {
+	// add the incoming hostname to the hostList before we sort and process it
+	hostList = append(hostList, hostname)
+
+	hostMap := make(map[string][]string)
+	// iterate over the full hostList
+	// if the element is a wildcard, add it to the list of wildcards
+	// if the element is not a wildcard, convert it to a wildcard and add any hostnames it matches to a map
+	for _, element := range hostList {
+		// FQDN-based hosts are grouped under a wildcard key
+		if hostnameContainsDot(element) {
+			wildcard := wildcardFromHostname(element)
+			if !slices.Contains(hostMap[wildcard], element) {
+				hostMap[wildcard] = append(hostMap[wildcard], element)
+			}
+		} else {
+			// any non-wildcard hosts go into the "extra" key and will be processed separately at the end
+			hostMap["extra"] = append(hostMap["extra"], element)
```

Comment:
Bit of a nit maybe, I think this comment is a little misleading? They're not really processed at the end so much as whenever we iterate over the "extra" element.

I'd also maybe argue we can just have a fully separate `extraHosts` list or something instead of a constant key coexisting in the map. I don't think collision is possible but we could probably reduce cognitive complexity here by keeping separate things separate.